rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,685 @@
+private key PEM, this file holds information necessary to retrieve the PIV
+handler for the hardware private key. Currently, this is just the YubiKey
+serial number and PIV slot.
+
+However, in order to use the key, the client must also know the public key,
+the private key policy to determine when to include touch/pin prompts, and the
+attestation statement to include in any re-login features (e.g. Per-session MFA).
+
+This additional information is retrieved directly from the hardware key by each
+call of the client process. This results in two problems:
+
+* Connecting to the hardware key and retrieving this information has some
+performance cost.
+  * In particular, re-attesting the key against the Yubico CA to derive the
+  private key policy takes upwards of 100ms.
+* Retrieving the information requires a mutually exclusive connection directly
+to the hardware key.
+  * The hardware key agent will hold open connections for at least 1 second in
+  order to improve performance for back-to-back signature requests, meaning
+  dependent clients could be locked out for this duration when trying to make
+  a direct connection.
+  * It would make more sense not to have dependent clients switching between
+  direct hardware key connections and the hardware key agent, especially from
+  an implementation perspective.
+
+Since all this information is known at time of login, we will instead store
+this information in the pem-encoded file:
+
+```diff
+-----BEGIN PIV YUBIKEY PRIVATE KEY-----
+######## PEM encoded #########
+{
+  "serial_number": "12345678",
+  "slot_key": "9e"
++ "public_key_der": "...",
++ "private_key_policy": "hardware_key_pin",
++ "attestation_statement": {...}
+}
+##############################
+-----END PIV YUBIKEY PRIVATE KEY-----
+```
+
+Note: expanding the PEM encoded file is the chosen solution as it also improves
+performance in the base, agentless case. For posterity, I originally planned on
+adding a hardware key agent rpc like `GetAdditionalInfo` for dependent clients
+to retrieve this info.
+
+Note: for backwards compatibility, clients will continue to retrieve this
+information directly from the hardware key if it is missing from the PEM
+encoded file.
+
+#### `hardwareKeyAgentService` pseudo-code implementation
+
+Teleport clients will use a `hardwareKeyAgentService` interface to interact
+with hardware private keys, rather than interacting directly with the PIV
+interface.
+
+Note: `HardwarePrivateKeyRef` is marshalled to/from the PEM encoded file described above.
+
+```go
+// Pseudo-code
+
+// hardwareKeyAgentService has two implementations:
+//  - direct implementation with piv-go, adapted slightly from our existing implementation
+//  - hardware key agent gRPC service implementation
+type hardwareKeyAgentService interface {
+  Sign(ref HardwarePrivateKeyRef, rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error)
```

Comment:
nit: I'd drop `rand` from the interface if possible, since you won't really be passing the rand source over gRPC. I think it's okay for the crypto.Signer implementation to ignore the `rand` argument

Reply:
The agent will use `rand` on the client side to generate the salt when needed, and the normal direct PIV implementation of the `HardwareKeyService` will use `rand` as normal.

Reply:
ah okay